File: flutter_app/lib/services/post_service.dart
import '../models/post.dart';
import 'api_service.dart';

/// Post/News service for managing posts
class PostService {
  final ApiService _api = ApiService();

  /// Get all posts
  Future<List<Post>> getAllPosts() async {
    final response = await _api.get('/api/posts');
    return (response as List).map((json) => Post.fromJson(json)).toList();
  }

  /// Get post by ID (increments view count)
  Future<Post> getPostById(int id) async {
    final response = await _api.get('/api/posts/$id');
    return Post.fromJson(response);
  }

  /// Get posts by category
  Future<List<Post>> getPostsByCategory(String category) async {
    final response = await _api.get('/api/posts/category/$category');
    return (response as List).map((json) => Post.fromJson(json)).toList();
  }

  /// Get published posts
  Future<List<Post>> getPublishedPosts() async {
    final response = await _api.get('/api/posts/published');
    return (response as List).map((json) => Post.fromJson(json)).toList();
  }

  /// Create post (admin only)
  Future<Post> createPost(Post post) async {
    final response = await _api.post('/api/posts', body: post.toJson());
    return Post.fromJson(response);
  }

  /// Update post (admin only)
  Future<Post> updatePost(int id, Post post) async {
    final response = await _api.put('/api/posts/$id', body: post.toJson());
    return Post.fromJson(response);
  }

  /// Delete post (admin only)
  Future<void> deletePost(int id) async {
    await _api.delete('/api/posts/$id');
  }
}

File: nodejs-backend/src/services/post.service.js
/**
 * Post Service
 * Handles post/news-related business logic
 */

const { Post } = require('../models');

class PostService {
  /**
   * Get all posts
   */
  async getAllPosts() {
    return await Post.findAll({
      order: [['date', 'DESC']]
    });
  }

  /**
   * Get post by ID
   */
  async getPostById(id) {
    return await Post.findByPk(id);
  }

  /**
   * Get posts by category
   */
  async getPostsByCategory(category) {
    return await Post.findAll({
      where: { category },
      order: [['date', 'DESC']]
    });
  }

  /**
   * Get published posts
   */
  async getPublishedPosts() {
    return await Post.findAll({
      where: { is_published: true },
      order: [['date', 'DESC']]
    });
  }

  /**
   * Create new post
   */
  async createPost(postData) {
    return await Post.create({
      title: postData.title,
      excerpt: postData.excerpt,
      description: postData.description,
      category: postData.category,
      author: postData.author,
      date: postData.date || new Date(),
      is_published: postData.isPublished || false,
      view_count: postData.viewCount || 0
    });
  }

  /**
   * Update post
   */
  async updatePost(id, postData) {
    const post = await Post.findByPk(id);
    if (!post) {
      throw new Error('Post not found');
    }

    if (postData.title !== undefined) post.title = postData.title;
    if (postData.excerpt !== undefined) post.excerpt = postData.excerpt;
    if (postData.description !== undefined) post.description = postData.description;
    if (postData.category !== undefined) post.category = postData.category;
    if (postData.author !== undefined) post.author = postData.author;
    if (postData.isPublished !== undefined) post.is_published = postData.isPublished;
    if (postData.viewCount !== undefined) post.view_count = postData.viewCount;

    await post.save();
    return post;
  }

  /**
   * Delete post
   */
  async deletePost(id) {
    const post = await Post.findByPk(id);
    if (!post) {
      throw new Error('Post not found');
    }

    await post.destroy();
    return true;
  }

  /**
   * Increment view count
   */
  async incrementViewCount(id) {
    const post = await Post.findByPk(id);
    if (post) {
      post.view_count = (post.view_count || 0) + 1;
      await post.save();
    }
  }
}

module.exports = new PostService();
